webserver/webserver/src/main/java/io/helidon/webserver/http1/Http1Connection.java

```diff
@@ -183,30 +187,35 @@ public void handle(Semaphore requestSemaphore) throws InterruptedException {
                                 }
                                 this.upgradeConnection = upgradeConnection;
                                 // this will block until the connection terminates
-                                upgradeConnection.handle(requestSemaphore);
+                                upgradeConnection.handle(limit);
                                 return;
                             }
                         }
                     }
                 }
-                if (requestSemaphore.tryAcquire()) {
-                    try {
-                        this.lastRequestTimestamp = DateTime.timestamp();
-                        route(prologue, headers);
-                        this.lastRequestTimestamp = DateTime.timestamp();
-                    } finally {
-                        requestSemaphore.release();
-                    }
-                } else {
+
+                Optional<LimitAlgorithm.Token> token = limit.tryAcquire();
+                if (token.isEmpty()) {
                     ctx.log(LOGGER, TRACE, "Too many concurrent requests, rejecting request and closing connection.");
                     throw RequestException.builder()
                             .setKeepAlive(false)
                             .status(Status.SERVICE_UNAVAILABLE_503)
                             .type(EventType.OTHER)
                             .message("Too Many Concurrent Requests")
                             .build();
-                }
+                } else {
+                    LimitAlgorithm.Token permit = token.get();
 
+                    try {
+                        this.lastRequestTimestamp = DateTime.timestamp();
+                        route(prologue, headers);
+                        permit.success();
+                        this.lastRequestTimestamp = DateTime.timestamp();
+                    } catch (Throwable e) {
+                        permit.dropped();
```

Comment:
Maybe something for a future PR: Should we allow a user to specify which exception can be recorded as ignored instead of being counted as dropped?